commandapi-core/src/main/java/dev/jorel/commandapi/arguments/ExceptionHandlingArgumentType.java

```diff
@@ -0,0 +1,69 @@
+package dev.jorel.commandapi.arguments;
+
+import com.mojang.brigadier.StringReader;
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+import dev.jorel.commandapi.exceptions.WrapperCommandSyntaxException;
+import dev.jorel.commandapi.wrappers.WrapperStringReader;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An {@link ArgumentType} that wraps another {@link ArgumentType} and intercepts any
+ * {@link CommandSyntaxException} to send to a developer-specified {@link InitialParseExceptionHandler}
+ *
+ * @param baseType The {@link ArgumentType} this object is wrapping.
+ * @param exceptionHandler The {@link InitialParseExceptionHandler} that handles intercepted {@link CommandSyntaxException}.
+ * @param exceptionParser A function that parses the information in a {@link CommandSyntaxException} to create an
+ * {@link ExceptionInformation} object.
+ *
+ * @param <T> The object returned when the wrapped {@link ArgumentType} is parsed.
+ * @param <ExceptionInformation> The class that holds information about the exception.
+ */
+public record ExceptionHandlingArgumentType<T, ExceptionInformation>(
```

Comment:
I think this name might not make the most sense. Although reading your essay (and hopefully understanding it well enough to write a good code review), once I saw this class a question that popped up in my head was why this is only for an `InitialParseExceptionHandler`. Upon reading the rest of the class it obviously made sense but I think if anyone crosses this class because they look through the code and didn't read the explanation in this PR before might also question the name.

I think it would make sense to call this something like `InitialParseExceptionHandlingArgumentType` or if you think that's too long, you can come up with another name? The important part for me would be including `InitialParse` somewhere in the name.

Reply:
Yeah, that makes sense. `InitialParseExceptionHandlingArgumentType` is long, but I think that's fine :P.

I'll also change the names of the [`InitialParseExceptionArgumentSerializer_(VERSION)`](https://github.com/JorelAli/CommandAPI/blob/73b2d122ed05cb6eaaa107bc789200ecf2cc89ee/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-nms/commandapi-bukkit-1.15/src/main/java/dev/jorel/commandapi/nms/ExceptionHandlingArgumentSerializer_1_15.java) and [`ExceptionHandlingArgumentInfo_(VERSION)`](https://github.com/JorelAli/CommandAPI/blob/73b2d122ed05cb6eaaa107bc789200ecf2cc89ee/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-nms/commandapi-bukkit-1.19-common/src/main/java/dev/jorel/commandapi/nms/ExceptionHandlingArgumentInfo_1_19_Common.java) classes.